### PR TITLE
Expose SystemMeta field accessors (#5497)

### PR DESCRIPTION
**Pull request by WinstonHartnett**
Adds user-facing immutable and unsafe mutable accessors to `SystemMeta` (see #5497).

### CODE REVIEW

**Review by mockersf — CHANGES_REQUESTED** [review]
the safety comments should give an indication on how to use them safely

File: crates/bevy_ecs/src/system/function_system.rs [hunk 1]
```diff
@@ -56,6 +86,18 @@ impl SystemMeta {
...
         self.is_send = false;
     }
+
+    /// Returns this system's last change tick.
+    #[inline]
+    pub fn last_change_tick(&self) -> u32 {
+        self.last_change_tick
+    }
+
+    /// Set this system's last change tick.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/function_system.rs:96]: Needs links and warnings.

File: crates/bevy_ecs/src/system/function_system.rs [hunk 2]
```diff
@@ -56,6 +86,18 @@ impl SystemMeta {
     pub fn set_non_send(&mut self) {
         self.is_send = false;
     }
+
+    /// Returns this system's last change tick.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/function_system.rs:90]: Needs links.

File: crates/bevy_ecs/src/system/function_system.rs [hunk 3]
```diff
@@ -43,6 +43,36 @@ impl SystemMeta {
         &self.name
     }
 
+    /// Returns the system's component access set.
+    #[inline]
+    pub fn component_access_set(&self) -> &FilteredAccessSet<ComponentId> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/function_system.rs:48]: These should all be `const`, along with the other methods on this type.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Wildly unsafe, but I agree that this should exist for advanced users to actually impl SystemParam manually. Needs more warnings, links and more helpful safety comments though.

File: crates/bevy_ecs/src/system/function_system.rs [hunk 4]
```diff
@@ -43,6 +43,36 @@ impl SystemMeta {
         &self.name
     }
 
+    /// Returns the system's component access set.
```
Comment (james7132) [inline crates/bevy_ecs/src/system/function_system.rs:46]:
```suggestion
    /// Fetches the system's component access set.
```

File: crates/bevy_ecs/src/system/function_system.rs [hunk 5]
```diff
@@ -43,6 +43,36 @@ impl SystemMeta {
...
+    #[inline]
+    pub fn component_access_set(&self) -> &FilteredAccessSet<ComponentId> {
+        &self.component_access_set
+    }
+
+    /// Returns a mutable reference to the system's component access set.
```
Comment (james7132) [inline crates/bevy_ecs/src/system/function_system.rs:52]:
```suggestion
    /// Fetches a mutable reference to the system's component access set.
```

File: crates/bevy_ecs/src/system/function_system.rs [hunk 6]
```diff
@@ -43,6 +43,36 @@ impl SystemMeta {
...
+    ///
+    /// # Safety
+    /// This allows unsafe modifications to the component access set that can violate Bevy's scheduling soundness.
```
Comment (james7132) [inline crates/bevy_ecs/src/system/function_system.rs:55]: This isn't documenting the actual safety invariant. Editing the component access set needs to accurately reflect the actual mutable and immutable accesses the system (or parameter) fetches from the World.

File: crates/bevy_ecs/src/system/function_system.rs [hunk 7]
```diff
@@ -43,6 +43,36 @@ impl SystemMeta {
...
+    /// Returns the system's archetype component access set.
+    #[inline]
+    pub fn archetype_component_access(&self) -> &Access<ArchetypeComponentId> {
+        &self.archetype_component_access
+    }
+
+    /// Returns a mutable reference to this system's archetype component access set.
+    ///
+    /// # Safety
+    /// This allows unsafe modifications to the archetype component access set that can violate Bevy's scheduling soundness.
```
Comment (james7132) [inline crates/bevy_ecs/src/system/function_system.rs:70]: This also must exactly reflect the safety invariants of using this.

File: crates/bevy_ecs/src/system/function_system.rs [hunk 8]
```diff
@@ -43,6 +43,36 @@ impl SystemMeta {
...
+    pub fn component_access_set(&self) -> &FilteredAccessSet<ComponentId> {
+        &self.component_access_set
+    }
+
+    /// Returns a mutable reference to the system's component access set.
+    ///
+    /// # Safety
+    /// This allows unsafe modifications to the component access set that can violate Bevy's scheduling soundness.
+    #[inline]
+    pub unsafe fn component_access_set_mut(&mut self) -> &mut FilteredAccessSet<ComponentId> {
```
Comment (james7132) [inline crates/bevy_ecs/src/system/function_system.rs:57]: The `pub(crate)` on the fields should be removed and the `bevy_ecs` uses of the access sets should use these accessors.

**Review by james7132 — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/system/function_system.rs [hunk 9]
```diff
@@ -15,13 +15,13 @@ use super::{In, IntoSystem, ReadOnlySystem};
 /// The metadata of a [`System`].
 #[derive(Clone)]
 pub struct SystemMeta {
```
Comment (mscofield0) [inline crates/bevy_ecs/src/system/function_system.rs:17]: When this gets rebased, please also make the spans public too.

**Comment by richchurcher** [comment]
Backlog cleanup: closing due to inactivity, noting existing tracking issue #5497.

**Pull request closed without merging**